src/shared/watchdog.c

```diff
@@ -114,9 +117,37 @@ int watchdog_set_timeout(usec_t *usec) {
         return r;
 }
 
+usec_t watchdog_runtime_wait(void) {
+        usec_t rtwait = USEC_INFINITY;
+        usec_t ntime;
+
+        if (timestamp_is_set(watchdog_timeout)) {
+                if (timestamp_is_set(watchdog_last_ping)) {
+                        ntime = now(clock_boottime_or_monotonic());
+                        assert(ntime >= watchdog_last_ping);
+                        rtwait = watchdog_last_ping + (watchdog_timeout / 2) - ntime;
+                } else
+                        rtwait = watchdog_timeout / 2;
+
+                if (rtwait <= 0)
```

Comment:
This means if we fail to ping the watchdog till half time we don't allow long sleeps and we keep trying to ping during our second half window with impact on CPU utilisation if the pings keep failing. This is different than previous behaviour when we always slept for half the watchdog_timeout no matter if the last ping was successful or not. In my opinion this is more robust, but I'm not sure if is acceptable so it needs to discussion.